[RUBY-3268] Add exemption handling and improve registration details display

jjromeo opened the pull request:
Jira ticket: https://eaflood.atlassian.net/browse/RUBY-3268

- Update `create_registration` to handle exemptions from params.
- Add `registration_exemptions_by_codes` method for dynamic exemptions.
- Rename `registration_exemptions` to `registration_exemptions_by_count`.
- Enhance `show.html.erb` to display detailed registration and exemption information.

Also includes some maintenance fixes for back-office

jjromeo committed:
[RUBY-3268] Add exemption handling and improve registration details display

- Update `create_registration` to handle exemptions from params.
- Add `registration_exemptions_by_codes` method for dynamic exemptions.
- Rename `registration_exemptions` to `registration_exemptions_by_count`.
- Enhance `show.html.erb` to display detailed registration and exemption information.

jjromeo committed:
[RUBY-3268] Refactor `TestingController` and enhance test coverage

- Refactor `create_registration` method for better readability
- Add test cases for various scenarios:
  - Registration creation with default exemptions
  - Registration creation with specified exemptions
  - Correct setting of expiry date
  - Population of `edit_token_created_at`

PaulDoyle-EA commented on app/controllers/testing_controller.rb:
```diff
@@ -8,22 +8,29 @@ class TestingController < ApplicationController
     registration.regenerate_and_timestamp_edit_token
 
-    render :show, locals: { registration: registration }
+    render :show, locals: { registration: registration, registration_exemptions: @registration_exemptions }
```
Minor: Shouldn't need to explicitly pass registration_exemptions to the view as can just access registration.registration_exemptions

jjromeo committed:
[RUBY-3268] Refactor `TestingController#create_registration` to use local variable `registration_exemptions` instead of instance variable.

jjromeo committed:
[RUBY-3268] rubocop fixes

jjromeo committed:
ReSonarCloud

jjromeo committed:
Merge branch 'main' into RUBY-3268-wex-extend-create-registration-end-point-to-include-renewal-token

jjromeo committed:
Re-add missing devise schema attributes

jjromeo committed:
Merge branch 'RUBY-3268-wex-extend-create-registration-end-point-to-include-renewal-token' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-3268-wex-extend-create-registration-end-point-to-include-renewal-token

jjromeo committed:
Remove ENV/env? mocking after specs

jjromeo committed:
[RUBY-3268] Refactor `registration_exemptions_by_codes` method to avoid creating non-existing exemptions and add test to ensure only existing exemptions are used

jjromeo committed:
[RUBY-3268] Refactor `TestingController` exemption methods and update specs

- Renamed methods `registration_exemptions_by_count` and `registration_exemptions_by_codes` to `create_registration_exemptions_by_count` and `create_registration_exemptions_by_codes` for clarity.
- Updated controller logic to use newly named methods and properly assign `registration_exemptions`.
- Simplified `create_registration_exemptions_by_codes` method by removing unnecessary `compact` call.
- Adjusted test to

jjromeo committed:
[RUBY-3268] Add `bucket_type` attribute to `bucket` factory and update `UpdateExemptionsService` spec to use specific `bucket_type`
